### DISCUSSION TITLE
Set session after custom email verification

### DISCUSSION THREAD

**Discussion by davshoward** [nextauthjs/next-auth#2537] [Help]
I've got Next Auth setup and configured (all working well) with the Credentials provider as we have a custom API to manage login.

However, I'm trying to find a way to automatically login a user when we return an object with user data and token (same one that is set in the authorize call:

```js
Providers.Credentials({
name: 'Login',
credentials: {
	username: { label: 'Email', type: 'text' },
	password: { label: 'Password', type: 'password' },
},
authorize: async (credentials) => {
	const res = await fetch(process.env.CREDENTIALS_URL, {
		method: 'POST',
		headers: {
			accept: 'text/plain',
			'Content-Type': 'application/json',
		},
		body: JSON.stringify(credentials),
	});

	const user = await res.json();
	if (user.result == 'Success') {
		return user;
	} else {
		throw new Error('Your email and password do not match');
	}
},
}),
```

Is this possible? If so, how?

Cheers

**Comment by balazsorban44** [comment]
If you mean to forward that data to the client, check out the jwt and session callbacks https://next-auth.js.org/configuration/callbacks

**Reply by davshoward** [reply]
Sorry for the delayed reply.

Yeah - so essentially, on a page pass a returned token to the session/jwt callbacks.

E.g. 

```
userService
    .verify()
    .then((response) => {
       // Set session/JWT to response
    })
    .catch((error) => {
        console.error('Error:', error);
    });
```

**Reply by balazsorban44** [reply]
Sorry, I am not sure I understand. Could you try to explain a bit better what you are trying to achieve?

**Reply by davshoward** [reply]
Sorry sure :-)

I'm sending a user to a /verify location with custom id to "verify" their email address, e.g. /verify?check=asdasdkaejrklwenfmsdnfkjsdhfks. Upon successful verification, the API returns an active/valid session as JSON (the same one we use to login). 

In my [...nextauth].tsx I have the following callbacks:

```
const callbacks = {
	async jwt(token, user) {
		if (user) {
			token.user = user;
		}

		return token;
	},

	async session(session, token) {
		if (token) {
			session.user = token.user.data;
			session.token = token.user.token;
		}

		return session;
	},
};
```

This works at login no problem, I just want to do the same again but now on my /verify page.

So in then success `.then` I want to be able to pass the returned session data to nextauth to then automatically login the user.

**Reply by balazsorban44** [reply]
We do have a similar provider to help with this: https://next-auth.js.org/providers/email would that help?
